File: hackerearth-vignan/III-SEM/challenge_18/02_binary_search_tree.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node *left, *right;
} Node;

Node *createNode(int data) {
    Node *newNode = malloc(sizeof(Node));
    newNode->data = data;
    newNode->left = newNode->right = NULL;
    return newNode;
}

Node *insertNode(Node *root, int val) {
    if (!root)
        return createNode(val);
    (val < root->data) ? (root->left = insertNode(root->left, val)) : (root->right = insertNode(root->right, val));
    return root;
}

void inorderTraverse(Node *root) {
    if (root) {
        inorderTraverse(root->left);
        printf("%d\n", root->data);
        inorderTraverse(root->right);
    }
}

int main() {
    int t;
    scanf("%d", &t);
    while (t--) {
        Node *root = NULL;
        int val;
        while (scanf("%d", &val) != EOF) {
            root = insertNode(root, val);
            if (getchar() == '\n')
                break;
        }
        inorderTraverse(root);
        printf("\n");
    }
    return 0;
}

File: hackerearth-vignan/III-SEM-2022-26/challenge_18/02_binary_search_tree.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


def create_node(data):
    newNode = Node(data)
    return newNode


def insert(root, item):
    if not root:
        return create_node(item)
    if item < root.data:
        root.left = insert(root.left, item)
    else:
        root.right = insert(root.right, item)
    return root


def inorder(root):
    if root:
        inorder(root.left)
        print(root.data)
        inorder(root.right)


def main():
    t = int(input())
    for _ in range(t):
        root = None
        nodes = list(map(int, input().split()))
        for item in nodes:
            root = insert(root, item)
        inorder(root)
        print()


if __name__ == "__main__":
    main()
